src/ucp/rndv/proto_rndv.c

```diff
@@ -720,13 +720,23 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_proto_rndv_send_reply,
     ucp_worker_cfg_index_t rkey_cfg_index;
     ucp_proto_select_param_t sel_param;
     ucp_proto_select_t *proto_select;
+    ucs_sys_device_t sys_dev;
     ucs_status_t status;
     ucp_rkey_h rkey;
 
     ucs_assert((op_id >= UCP_OP_ID_RNDV_FIRST) &&
                (op_id < UCP_OP_ID_RNDV_LAST));
 
     if (rkey_length > 0) {
+        /* Sys device is crucial for rkey_unpack in rndv pipeline protocols that
+         * use CUDA staging buffers. If sys_dev of the user memory is unknown
+         * (e.g., host or CUDA-managed memory), use the default sys_dev since it
+         * will be used for allocating staging buffers anyway. */
+        sys_dev =
+        (req->send.state.dt_iter.mem_info.sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN) ?
```

Comment:
Maybe it is better to include the condition to the method, and  use the comment as a description of the method.

Reply:
imo it would be harder to read in rndv code, as now we check explicitly whether we need to use default sys device id or the one from user buffer

Reply:
Or at least this:
```C
        sys_dev = req->send.state.dt_iter.mem_info.sys_dev;
        if (sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN) {
            sys_dev = ucp_context_default_sys_dev(worker->context, UCS_MEMORY_TYPE_CUDA);
        }
```

Reply:
ok

Reply:
force-pushed because PR is small